### ISSUE TITLE
Emulator crash (Access Violation) in Wario Land II (U)

### ISSUE THREAD

**Issue by Raijinili** [mgba-emu/mgba#989]
- mGBA 0.6.1
- Windows 10

After clearing a stage and starting another stage, mGBA crashes. It happens pretty consistently, but sometimes doesn't happen. I don't know why.

Save state attached. Press A to start the stage and cause the crash.

![wario land ii u ss8](https://user-images.githubusercontent.com/2422807/36060686-55e9741a-0e1c-11e8-9098-51b06675fa1d.png)

**Comment by rootfather** [comment]
Where's the save state? All I can find in this post is a screenshot...

**Comment by Raijinili** [comment]
Screenshot is the save state.

I'm AFK at the moment, so I can't say.

**Comment by endrift** [comment]
Seems to work fine for me. There were some bugs regarding savestates and this game that should have been fixed now though.

**Comment by Cibomatto2002** [comment]
It happens at random. If you play the game for a long time I will crash as random places.

**Comment by Raijinili** [comment]
@Cibomatto2002 That's... less informative than my original report. What version of mGBA are you using?

**Comment by Raijinili** [comment]
@rootfather Still crashing on mGBA-build-2018-02-07-win32-5030-eafd265ab6b0b025279dea5d6a485ecaa6b724a9.

@endrift Yeah, it's the savestate. I just played through a level on the above build without using savestates, saved state before starting the next level, entered it without a problem, reloaded the save, and it crashed.

**Comment by endrift** [comment]
Yeah, I can't reproduce this on latest. Can I see your config.ini?

**Comment by Raijinili** [comment]
```ini
[ports.qt]
gbc.bios=
preload=0
width=720
useBios=0
audioBuffers=1536
lockAspectRatio=1
volume=256
shader=
rewindSave=1
savestatePath=
screenshotPath=
libraryStyle=0
fastForwardRatio=-1
sampleRate=44100
mute=0
rewindBufferCapacity=300
allowOpposingDirections=0
suspendScreensaver=1
savegamePath=
patchPath=
idleOptimization=ignore
scaleMultiplier=3
lastDirectory=<REDACTED>
gba.bios=
audioSync=1
frameskip=0
fpsTarget=60
skipBios=0
rewindEnable=0
loadStateExtdata=9
gb.bios=
videoSync=0
pauseOnFocusLost=0
showLibrary=0
height=480
fullscreen=0
lockIntegerScaling=1
resampleVideo=0
saveStateExtdata=31

[gba.input-profile.Logitech(R) Precision(TM) Gamep]
keyR=5
keyL=4
axisUpAxis=-1
hat0Right=4
gyroSensitivity=2.2e+009
keyStart=9
keyLeft=-1
keyDown=-1
hat0Left=5
axisLeftAxis=-0
keyB=1
keyUp=-1
hat0Down=7
axisRightValue=12288
axisRightAxis=+0
axisUpValue=-12288
axisDownAxis=+1
gyroAxisX=0
tiltAxisX=2
axisDownValue=12288
keyA=2
keyRight=-1
hat0Up=6
gyroAxisY=1
axisLeftValue=-12288
keySelect=8
tiltAxisY=3

[gba.input-profile.Logitech(R) Precision(TM) Gamepa”ÒB]
gyroSensitivity=2.2e+009
tiltAxisX=2
gyroAxisX=0
gyroAxisY=1
tiltAxisY=3

[gba.input.SDLB]
keyR=5
keyL=4
axisUpAxis=-1
hat0Right=4
gyroSensitivity=2.2e+009
keyStart=9
keyLeft=-1
keyDown=-1
hat0Left=5
axisLeftAxis=-0
keyB=1
keyUp=-1
hat0Down=7
axisRightValue=12288
axisRightAxis=+0
axisUpValue=-12288
axisDownAxis=+1
gyroAxisX=0
tiltAxisX=2
axisDownValue=12288
keyA=2
keyRight=-1
device0=Logitech(R) Precision(TM) Gamepad
hat0Up=6
gyroAxisY=1
axisLeftValue=-12288
keySelect=8
tiltAxisY=3

[gba.input-profile.Logitech(R) Precision(TM) Gamepad]
keyR=5
keyL=4
axisUpAxis=-1
hat0Right=4
gyroSensitivity=2.2e+009
keyStart=9
keyLeft=-1
keyDown=-1
axisLeftAxis=-0
hat0Left=5
keyB=1
keyUp=-1
axisRightValue=12288
axisRightAxis=+0
axisUpValue=-12288
axisDownAxis=+1
hat0Down=7
axisDownValue=12288
tiltAxisX=2
gyroAxisX=0
keyA=2
keyRight=-1
axisLeftValue=-12288
hat0Up=6
gyroAxisY=1
keySelect=8
tiltAxisY=3

[gba.input.QT_K]
keyR=83
keyL=65
keyStart=16777220
keyLeft=16777234
keyDown=16777237
keyB=90
keyUp=16777235
keyA=88
keyRight=16777236
keySelect=16777219
```

**Comment by Raijinili** [comment]
Anything to do on my end to help track down the problem? I'm not especially familiar with debuggers. I do have some experience and theoretical knowledge about them, but I don't know how to start the debugging process for Windows applications. How do I get the debugger to tie running code to source code? Do I use a special version of mGBA? Do I need to compile it myself?

**Comment by endrift** [comment]
Is this still a problem in the latest dev builds?

**Comment by Raijinili** [comment]
Using latest: `mGBA-build-2018-09-02-win32-5211-7d79db7d7d31f5a956123e751f455eee9d4b043e`

The first time I opened it, it crashed after "Continue".

The second time, it crashed after I entered a stage and idled for a minute (while not focused).

The third time, it crashed after idling on the starting menu (not focused).

The fourth time, it lasted long enough in idle (still not focused) to play the cutscene. I skipped the cutscene, let it idle again (still background), and it crashed.

The fifth time, I opened the game, didn't press any buttons, and let it idle. It lasted a minute or two (started to play a cutscene) and crashed.

The sixth and seventh time, I decided to stop testing the new crash, loaded an old savestate, and started a new stage. It crashed. The original issue is still a problem.

**Comment by endrift** [comment]
*May* be fixed in 880a86df301a93e60bf6e07eb08e5c418a71f09d, but I'm doubtful. I still haven't really been able to reproduce, and ASAN can't find it either.

**Comment by Raijinili** [comment]
Not fixed.

The crash started occurring in build 4776, commit c030fb65.

**Comment by endrift** [comment]
Thanks for the info! I'll take a closer look in a bit.

**Comment by endrift** [comment]
Possibly fixed by 69c2969db01c900d000e9c7cdba83db0fd8036e2

**Comment by Raijinili** [comment]
Still crashing on build 5248-69c2969.

**Comment by endrift** [comment]
I'm kind of out of ideas. Is there anything specific about your setup that might be relevant, like the ROM being on a network share, or a read-only directory?

**Comment by Raijinili** [comment]
No. I've made progress in the game while having the issue.

It should be an in-memory issue, because it only occurs by a particular
in-game action, not during load.

When I can, I'll test if it's a game memory issue:
1. Save with savedata
2. Load without savedata. (Corruption may occur here.)
3  Save with savedata again. (Save the corruption.)
4. Load with savedata. (Normally fine, but now loads the corruption.)
5. Start level.
6. If it crashes, restart and repeat 4-5.

If it crashes at 5 and 6, it may be a game state corruption. If it doesn't
crash at 5, or crashes at 5 but not 6, it is definitely an emulator state
corruption.

I will also try to see if the corruption persists through another save/load
cycle.
1. Save with savedata.
2. Load without savedata.
3. Save without savedata.
4. Load without savedata.
5. Start.

I suspect the problem is emulator, not game, state corruption. I'll look at
the source for what happens when it tries to ignore existent savedata.

I can also try to step through the mGBA debugger for the crash. I bet it's
a device read, maybe an SRAM read.

Can you teach me how to build with debugging symbols? Should the nightlies
have debugging symbols by default?

**Comment by endrift** [comment]
Build instructions are provided in the README I believe. I want to do some magic to provide the symbols as a separate download because they are big, but I haven't gotten to it yet. I may look into it later today if that helps.

**Comment by Raijinili** [comment]
Test 1: Crashed on step 5. Did not crash on step 6. Therefore, the emulator's runtime state, not the saved state, has the corruption.

Test 2: Crashed. I don't think it tells us anything new.

Test 3: Testing state resets. After setting up the corruption, I do one of the following, then start a stage.
- Load another game, then load Wario Land 2: **OK**
- Load Wario Land 2 again: **OK**
- Reset (Ctrl+R): **OK**
- GBC restart (A+B+Start+Select): **Crash**

---

Test 4: Debugger. I'm just stumbling through this.
1. `break 1` seems to pause execution.
2. Then I go to the game window and hold down the "A" button while clicking the debugger window.
3. Repeatedly execute `frame`, counting the number of times I executed it. With the attached state, the crash happens on the twelfth advance.
![wario land ii u ss1](https://user-images.githubusercontent.com/2422807/45602298-ebd62a80-b9e9-11e8-9ebd-ff84ab837754.png)

I tried to save a state (through the "File" menu) right before the crash advance, but it caused mGBA (barely any CPU use). Maybe it's because you can't save/load states that way while debugging. I don't want to make another issue right now to narrow it down, but `save 9` worked. The created state follows:
![wario land ii u ss9](https://user-images.githubusercontent.com/2422807/45602400-723f3c00-b9eb-11e8-85bd-f3524d1ab270.png)

However, loading the state normally left me with a black screen, but the emulator was responsive and showed ~65.5 FPS. Using `load 9` through the debugger AFTER an explicit `break`, then using `frame` again, caused a hang with 2/4 cores monopolized, rather than a crash. Loading it during normal gameplay, and then using `frame` (because loading through the debugger halts execution anyway) sometimes causes the crash, sometimes the infloop, and sometimes the responsive black screen (possibly depending on what states I load before the final `load 9`). Yet another issue to flesh out later.

Anyway, then I tried to use that state and advance instructions to cause the crash. However, I greatly underestimated the number of instructions executed per frame, and I'm stuck in a loop of 1AB6-1AB7-1AB8. I can't figure out the syntax to set a breakpoint at `PC = 1ABA`, so I'll give up for now.

**Comment by endrift** [comment]
I think this may finally be fixed in adbda8510792919e82a89a9d94902b8d66d4f8fd

**Comment by Raijinili** [comment]
Still there. Tested on both an earlier state and a fresh new state.

**Comment by endrift** [comment]
Ok I'm like 80% sure it's actually fixed in 0d25e8079172aee82958ce4babf3e0ef470200fb.

**Comment by Raijinili** [comment]
It works!

I was making notes from more testing, and it feels like a waste to delete them.

> In the log, the following line comes up one frame before the crash, or right as the crash happens:
> 
>     [INFO] GB Memory:	Savedata synced
> 
> On another run, I notice that it says "Loading savedata" despite that option being off.
> 
>     [INFO] Savestate:	Loading RTC
>     [INFO] Savestate:	Loading screenshot
>     [INFO] Savestate:	Loading savedata
>     [INFO] Savestate:	Loading RTC
>     [INFO] Status:	State 1 loaded
> 
> On a normal run without loading state, the savedata sync message happens both before and after the critical point of stage entry. Two more copies get spit out a few seconds later.

**Comment by endrift** [comment]
Oh thank fuck it's finally fixed.

**Comment by Raijinili** [comment]
Any clue why you couldn't reproduce the issue?

**Comment by endrift** [comment]
Not really. That new state sometimes crashed and sometimes didn't though, so that finally helped me track it down. And I squashed some related bugs in the process, so that's a plus.
